### PR TITLE
pkp/pkp-lib#3594 Add entity schema, Context API, and Vue.js forms

### PR DESCRIPTION
**Pull request by NateWr**
- Add journal-specific context schema
- Update journal_settings db schema to remove setting_type
- Convert journal contact/masthead forms to Vue.js

### CODE REVIEW

File: dbscripts/xml/upgrade/3.2.0_archiving_settings.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+<!--
+  * dbscripts/xml/upgrade/3.2.0_archiving_settings.xml
+  *
+  * Copyright (c) 2013-2018 Simon Fraser University
+  * Copyright (c) 2003-2018 John Willinsky
+  * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+  *
+  * Remove unused context settings related to archiving plugins. These settings
+	* were never used, but were saved to the database as part of ArchivingForm's
```
Comment (asmecher) [inline dbscripts/xml/upgrade/3.2.0_archiving_settings.xml]: (should be space indents here)

File: dbscripts/xml/upgrade/3.2.0_setting_type.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+<!--
+  * dbscripts/xml/upgrade/3.2.0_setting_type.xml
+  *
+  * Copyright (c) 2013-2018 Simon Fraser University
+  * Copyright (c) 2003-2018 John Willinsky
+  * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+  *
+  * Remove the setting_type column from the journal_settings and site_settings
+	* tables. This column is obsolete now that we use the context.json schema to
```
Comment (asmecher) [inline dbscripts/xml/upgrade/3.2.0_setting_type.xml]: (indents again)

File: dbscripts/xml/upgrade/3.2.0_stylesheet.xml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  * Update the context setting `stylesheet` format to drop the image attributes
+	* object and replace it with a string referencing the filename. Also remove
+	* the unused original_style_file_name column from the site table.
+  *
+  -->
+
+<data>
+	<sql>
+		<query>
+			UPDATE journal_settings SET setting_value="styleSheet.css" WHERE setting_name="styleSheet";
```
Comment (asmecher) [inline dbscripts/xml/upgrade/3.2.0_stylesheet.xml]: SQL quotes should always be `'` (`"` breaks PostgreSQL)

File: js/controllers/grid/articleGalleys/ArticleGalleyGridHandler.js [hunk 4]
```diff
@@ -58,11 +58,18 @@
...
-		$.when($(selector)).then(function() {
-			$(function() {
+
+		function defferedClick(selector) {
+			if ($(selector).length) {
 				$(selector).click();
-			});
-		});
+			} else {
+				setTimeout(function() {
```
Comment (asmecher) [inline js/controllers/grid/articleGalleys/ArticleGalleyGridHandler.js]: I think this removes the example I cited earlier on the pkp-lib PR :) I liked the when/then formulation better; is there something here that breaks it?

**Comment by asmecher** [comment]
Same business here as I posted on the pkp-lib PR :)

**Comment by NateWr** [comment]
@asmecher When running the tests locally, I'm getting a bunch of failures in `MetadataPropertyTest`.

```
There were 11 failures:

1) PKPComponentRouterTest::testGetCacheFilename
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

2) FormValidatorLocaleEmailTest::testIsValid
Failed asserting that true is false.

/home/nate/Documents/ojs/lib/pkp/tests/classes/form/validation/FormValidatorLocaleEmailTest.php:35

3) FormValidatorUrlTest::testIsValid
Failed asserting that true is false.

/home/nate/Documents/ojs/lib/pkp/tests/classes/form/validation/FormValidatorUrlTest.php:41

4) DateStringNormalizerFilterTest::testExecute
Failed asserting that null matches expected '2003'.

/home/nate/Documents/ojs/lib/pkp/tests/classes/metadata/DateStringNormalizerFilterTest.php:28

5) MetadataPropertyTest::testCompositeWithoutParameter
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

6) MetadataPropertyTest::testCompositeWithWrongParameter
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

7) MetadataPropertyTest::testControlledVocabWithoutParameter
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

8) MetadataPropertyTest::testControlledVocabWithWrongParameter
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

9) MetadataPropertyTest::testNonParameterizedTypeWithParameter
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

10) MetadataPropertyTest::testSetUnsupportedType
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

11) MetadataPropertyTest::testSetUnsupportedCardinality
Failed asserting that exception of type "PHPUnit_Framework_Error" is thrown.

--

There were 12 skipped tests:

1) FileCacheTest::testCacheMiss

/home/nate/Documents/ojs/lib/pkp/tests/classes/cache/FileCacheTest.php:68

2) ConfigTest::testReloadDataAndGetData

/home/nate/Documents/ojs/lib/pkp/tests/classes/config/ConfigTest.php:53

3) ConfigTest::testGetVar
This test depends on "ConfigTest::testReloadDataAndGetData" to pass.

4) ConfigTest::testGetVarFromOtherConfig
This test depends on "ConfigTest::testGetVar" to pass.

5) PKPComponentRouterTest::testSupportsWithPathinfoUnsuccessfulComponentFileDoesNotExist

/home/nate/Documents/ojs/lib/pkp/tests/classes/core/PKPComponentRouterTest.php:91

6) PKPComponentRouterTest::testIsCacheable

/home/nate/Documents/ojs/lib/pkp/tests/classes/core/PKPRouterTestCase.inc.php:76

7) PKPPageRouterTest::testIsCacheable

/home/nate/Documents/ojs/lib/pkp/tests/classes/core/PKPRouterTestCase.inc.php:76

8) DBConnectionTest::testInitPostgresDBConnection

/home/nate/Documents/ojs/lib/pkp/tests/classes/db/DBConnectionTest.php:56

9) DBConnectionTest::testInitCustomDBConnection

/home/nate/Documents/ojs/lib/pkp/tests/classes/db/DBConnectionTest.php:71

10) FilterDAOTest::testCompositeFilterCrud

/home/nate/Documents/ojs/lib/pkp/tests/classes/filter/FilterDAOTest.php:96

11) FormValidatorPostTest::testIsValid
Disabled for static invocation of Request.

/home/nate/Documents/ojs/lib/pkp/tests/classes/form/validation/FormValidatorPostTest.php:34

12) XMLTypeDescriptionTest::testInstantiateAndCheck

/home/nate/Documents/ojs/lib/pkp/tests/classes/xslt/XMLTypeDescriptionTest.php:33

FAILURES!
Tests: 217, Assertions: 731, Failures: 11, Skipped: 12.
```

I'm at a bit of a loss. I'll dig in deeper tomorrow now that I know how to run a single PHPUnit test. But do you have any hints looking at those failures? I've never dealt with the `Filter` and `MetadataProperty` classes. I presume that the failures are due to changes in the `Validator` classes, but I can't quite see how.

Also, any idea what causes a test to be skipped?

**Comment by asmecher** [comment]
You might have to specify the path to the `phpunit` binary explicitly -- it should be the right version of PHPUnit as per the Composer dependencies (e.g. `lib/pkp/lib/vendor/bin/phpunit`).

**Comment by asmecher** [comment]
Tests are skipped when `$this->markTestSkipped()` is called in the test (not sure if there are other causes) -- there are some of these in the codebase, due to ill-maintained code.

**Comment by NateWr** [comment]
Just a few outstanding comments and that `DateStringNormalizerFilter` thing, and :crossed_fingers: I think we might be close to merge! Thanks for all the help so far!

File: classes/components/forms/context/ContextForm.inc.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	 * @copydoc PKPContextForm::__construct()
+	 */
+	public function __construct($action, $successMessage, $locales, $baseUrl, $context) {
+		parent::__construct($action, $successMessage, $locales, $baseUrl, $context);
+
+		$this->addField(new FieldText('abbreviation', [
+				'label' => __('manager.setup.journalAbbreviation'),
+				'isMultilingual' => true,
+				'value' => $context ? $context->getData('abbreviation') : null,
+			]), ['after', 'acronym']);
```
Comment (asmecher) [inline classes/components/forms/context/ContextForm.inc.php]: Should e.g. `'after'` be a defined constant?

Comment (NateWr) [reply classes/components/forms/context/ContextForm.inc.php]: :+1:

**Pull request merged**